### ISSUE TITLE
feature request: emulate the PS-key press

### ISSUE THREAD

**Issue by alexeykorobov** [dhleong/ps4-waker#10]
Would it be possible to emulate the PS-key press with the ps4-waker? It's a crucial functionality when interacting with games and apps, e.g. to be able to exit them. I see the PlayStation App has following keys:
1. PS-key (both short and long presses work)
2. Back
3. Arrow keys (by swiping)
4. OK (tap)
5. Options

**Comment by dhleong** [comment]
I believe it should be, but I haven't had time to look into it myself. I think someone was looking into it at one point, but I haven't heard back in a while....

**Comment by alexeykorobov** [comment]
Could you provide me a basic rundown on how to approach the task and what tools/tricks to use to capture the PS-key command and debug new code? I'd then try my luck with that :)

**Comment by dhleong** [comment]
The best way is probably to decompile the Playstation App, or another app that works with the Playstation SDK as a companion app (it seems they include the the full SDK). I started with the Playstation App for most of the features here, but this was pretty tough because it was heavily obfuscated. I hesitate to list alternative apps here, but the one I used for a few more recent features didn't have the same amount of obfuscation, so that helped. 

From there it's a matter of figuring out what classes implement the appropriate packets, and in what sequence they should be sent. Unfortunately, the the socket traffic is encrypted so it's somewhat difficult to capture directly. You can get what the client is sending, because we can extract the playstation's public key and the app's private key from the app, but there's no way to fake the responses to the app, since we don't know the playstation's private key. 

Long story short... there's a lot of educated guessing involved.

Good luck!

**Comment by alexeykorobov** [comment]
well, I didn't expect the decompile part :)

> You can get what the client is sending, because we can extract the playstation's public key and the app's private key from the app

How can I use the ps4-waker to capture what the PlayStation App is sending? I assume it has to be connected to the app in a fake PS mode, like it does in the initial step to record the credentials.

**Comment by dhleong** [comment]
> I assume it has to be connected to the app in a fake PS mode, like it does in the initial step to record the credentials.

Yep, that's the idea. But, there's no way to activate this directly in the code because it's not very useful—we can't send encrypted responses, so the client quickly gets confused. It was only really useful for figuring out the login sequence—for the other features of the lib, I had to look at the decompiled source/bytecode.

**Comment by dhleong** [comment]
Hi @alexeykorobov, not sure how far you got on this but I just did some experimenting and pushed the `dhleong/RemoteControl` branch. Feel free to play around with it and let me know how it works for you :)

See 295e99328baec38314f71fc7f779e462e87d1ac2

**Comment by Marcoaferrer** [comment]
First off

**Comment by Marcoaferrer** [comment]
First off I wanted to say thanks for all the hard work youve put into this lib. I was wondering if you made any progress on implementing methods to call remote commands. If not no worries I dont mind implementing it my self I just didnt want to fork the branch if youve already started on it. Ive been implementing Xbox Kinect style voice commands on my ps4 and this lib has done wonders for me.

**Comment by dhleong** [comment]
What commands specifically? The branch linked above works to send the `Socket.RCKeys` key presses, the topic of this issue. I think I never merged it because I was hoping for some feedback from someone who'll use it, and then just forgot about it. I'd be happy to merge and publish if it's working for you.

**Comment by Marcoaferrer** [comment]
I havent tried implementing ps remote commands yet I wanted to get in touch with you first. This past weekend I was able to set up voice commands for turning on / off the playstation as well as a global on / off with an ir emitter I made from an arduino. I also was able to add voice commands for starting titles and changing tv inputs.

**Comment by dhleong** [comment]
Gotcha, so you're just checking the status of this branch then? 

It looks like the cli interface is only half complete, so I'll try to finish that up and get it merged sometime this week.

Good luck with your project! I'm using my Echo to do something similar with this lib.

**Comment by Marcoaferrer** [comment]
Awesome thanks for all the help. I look forward to testing it out.

**Comment by dhleong** [comment]
Just published v0.10.0 with this support baked in. If you're looking to send key presses programmatically rather than through the cli, note that the PS4 appears to handle the presses on a best-effort basis and does not queue them up, so I recommend putting a delay between sends. The `ps4socket` is a low-level abstraction and does not handle this for you. There may be a day when a higher-level abstraction is built on top of that (and on top of which the cli can sit), but it probably won't be any time soon....
